Change: Groups cache vehicle lists

SamuXarick opened the pull request:
## Motivation / Problem
AIs creating lists of vehicles will slow down OpenTTD especially with high amount of vehicles in the game.
<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description
Create a `vehicle_list` member for `GroupStatistics`. Every group, including the `group_default` and the `group_all`, will keep track of `const Vehicle *`'s as they're added or removed from them.
<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations
This is a work in progress.

1. I'm looking into the rest of OpenTTD code for places where the use of this cache could be beneficial, but...
2. ... `AfterLoad` is the biggest hindrance to this approach. Some of its routines can at times call long established functions that are now being switched away from `VehicleListPool` iteration to the cached group vehicle list version. I am experimenting populating `GroupStatistics`, which is where these lists reside, as early as possible in the `AfterLoadGame` function.

The following have been changed so far:
- Iterate group vehicle lists for `GenerateVehicleSortList`.
- Iterate group vehicle lists for `ScriptVehicleList_Station`, `ScriptVehicleList_Depot`, `ScriptVehicleList_Group` and `ScriptVehicleList_DefaultGroup`.
- Iterate aircraft group list for `UpdateAirplanesOnNewStation`.
- Iterate group vehicle lists for `DisasterTick_Ufo` and `DisasterTick_Big_Ufo`. As a compromise, the result of this change differs from the original. The iteration order changed from lowest to highest vehicle index to lowest to highest company index plus the order at which each vehicle was added to each list regardless of their indexes.
- Iterate group vehicle lists for `UpdateCompanyRatingAndValue`.
- Iterate group vehicle lists for `SettingsDisableElrail`.
- Iterate group vehicle lists for `GetPreviewCompany` and `NewVehicleAvailable`. As a compromise, the result of this change differs from the original. Since the list only contains primary vehicles, free wagons aren't taken into consideration.
- Iterate group vehicle lists for `GroupStatistics::UpdateProfits`.
- Iterate group vehicle list for `PropagateChildLivery`.
- Iterate group vehicle list for `CmdAddSharedVehicleGroup`.
- Iterate group vehicle list for `CmdRemoveAllVehiclesGroup`.
- Iterate group vehicle lists for `WhoCanServiceIndustry`.
- Iterate group vehicle lists for `NetworkPopulateCompanyStats` and `NetworkAutoCleanCompanies`.
- Iterate group vehicle lists for `RemoveOrderFromAllVehicles`. Additionally make the function receive one parameter more with the owner of the destination, to help it perform faster.
<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

SamuXarick committed:
Change: Groups cache vehicle lists

Use std::vector<const Vehicle *> VehicleList for these lists.

SamuXarick committed:
Codechange: Iterate group vehicle lists for GenerateVehicleSortList

SamuXarick committed:
Codechange: Iterate group vehicle lists for ScriptVehicleList_Station, ScriptVehicleList_Depot, ScriptVehicleList_Group and ScriptVehicleList_DefaultGroup

glx22 commented on src/vehiclelist.cpp:
```diff
@@ -115,9 +115,13 @@ bool GenerateVehicleSortList(VehicleList *list, const VehicleListIdentifier &vli
-		case VL_STATION_LIST:
-			for (const Vehicle *v : Vehicle::Iterate()) {
-				if (v->type == vli.vtype && v->IsPrimaryVehicle()) {
+		case VL_STATION_LIST: {
```
`{` not needed

glx22 commented on src/vehiclelist.cpp:
```diff
@@ -128,6 +132,7 @@ bool GenerateVehicleSortList(VehicleList *list, const VehicleListIdentifier &vli
 				}
 			}
 			break;
+		}
```
`}` not needed

glx22 commented on src/vehiclelist.cpp:
```diff
@@ -140,38 +145,49 @@ bool GenerateVehicleSortList(VehicleList *list, const VehicleListIdentifier &vli
-					if (v->type == vli.vtype && v->IsPrimaryVehicle() &&
-							v->owner == vli.company && GroupIsInGroup(v->group_id, vli.index)) {
-						list->push_back(v);
+		case VL_GROUP_LIST: {
```
`{` not needed

glx22 commented on src/vehiclelist.cpp:
```diff
@@ -140,38 +145,49 @@ bool GenerateVehicleSortList(VehicleList *list, const VehicleListIdentifier &vli
+				}
+				break;
+
+			} else if (vli.index != ALL_GROUP) {
```
You can drop the `else` as there's a `break` in previous `if`

glx22 commented on src/vehiclelist.cpp:
```diff
@@ -140,38 +145,49 @@ bool GenerateVehicleSortList(VehicleList *list, const VehicleListIdentifier &vli
 				break;
 			}
 			FALLTHROUGH;
+		}
```
`}` not needed

SamuXarick committed:
Codechange: Iterate aircraft group list for UpdateAirplanesOnNewStation

SamuXarick committed:
Change: Iterate group vehicle lists for DisasterTick_Ufo and DisasterTick_Big_Ufo

As a compromise, the result of this change differs from the original. The iteration order changed from lowest to highest vehicle index to lowest to highest company index plus the order at which each vehicle was added to each list regardless of their indexes.

SamuXarick committed:
Codechange: Iterate group vehicle lists for UpdateCompanyRatingAndValue

SamuXarick committed:
Change: Iterate group vehicle lists for GetPreviewCompany and NewVehicleAvailable

As a compromise, the result of this change differs from the original. Since the list only contains primary vehicles, free wagons aren't taken into consideration.

nielsmh commented on src/elrail.cpp:
```diff
@@ -625,10 +625,12 @@ void SettingsDisableElrail(int32_t new_value)
+		/* Fix the total power and acceleration for trains */
+		for (const Vehicle *v : stats.vehicle_list) {
+			Train *t = Train::From(Vehicle::Get(v->index));
+			/* power and acceleration is cached only for front engines */
```
This comment is no longer relevant, since it related to the original following line, that tests for `IsFrontEngine()`. But since your collection now only contains front engines, it no longer relates to this. The more appropriate location might be to move it outside the loop and note that the power/acceleration values are only cached for front engines and that is what this collection enumerates.

nielsmh commented on src/vehiclelist.cpp:
```diff
@@ -116,8 +116,11 @@ bool GenerateVehicleSortList(VehicleList *list, const VehicleListIdentifier &vli
-				if (v->type == vli.vtype && v->IsPrimaryVehicle()) {
+			for (const Company *c : Company::Iterate()) {
+				/* Only iterate over all companies in the case of neutral Oil Rig stations */
+				if (vli.vtype != VEH_AIRCRAFT && vli.vtype != VEH_SHIP && vli.company != c->index) continue;
```
I'm not sure if this is correct. Shouldn't it rather test for the station owner being the neutral company, and not care about the vehicle type in question at all?

I also somewhat doubt this is going to be a performance improvement in this specific case. On the other hand, it's "only" a GUI use and not performance critical for the simulation itself.

SamuXarick replied:
Fixed. Thanks.

There's still a positive gain. In the worst case, it iterates only one vehicle type per company.

nielsmh reviewed (commented):
I believe the biggest gain for actual game performance would be splitting out the vehicle types for the core game loop. There already is (or has been) a PR for that but it has not been merged.

SamuXarick committed:
Codechange: Iterate group vehicle lists for GroupStatistics::UpdateProfits

SamuXarick committed:
Codechange: Iterate group vehicle list for PropagateChildLivery

SamuXarick committed:
Codechange: Iterate group vehicle list for CmdAddSharedVehicleGroup

SamuXarick committed:
Codechange: Iterate group vehicle list for CmdRemoveAllVehiclesGroup

SamuXarick committed:
Codechange: Iterate group vehicle lists for WhoCanServiceIndustry

SamuXarick committed:
Codechange: Iterate group vehicle lists for NetworkPopulateCompanyStats and NetworkAutoCleanCompanies

SamuXarick committed:
Codechange: Iterate group vehicle lists for RemoveOrderFromAllVehicles

Additionally make the function receive one parameter more with the owner of the destination, to help it perform faster.

SamuXarick committed:
Change: Update group statistics earlier in AfterLoad

This commit prefaces the next incoming set of commit changes in order to have the vehicle lists already generated as an attempt to prevent AfterLoad actions accessing them empty.

SamuXarick committed:
Codechange: Iterate group vehicle lists for SettingsDisableElrail

SamuXarick committed:
Codechange: Iterate group vehicle list for ScriptGroup::GetProfitThisYear and ScriptGroup::GetCurrentUsage

SamuXarick committed:
Codechange: Iterate group vehicle list for UpdateAllServiceInterval, UpdateServiceInterval, TrainAccelerationModelChanged, TrainSlopeSteepnessChanged, RoadVehAccelerationModelChanged, RoadVehSlopeSteepnessChanged, CheckFreeformEdges and InvalidateShipPathCache

SamuXarick committed:
Change: Iterate group vehicle lists for Station::~Station

As a compromise, the result of this change differs from the original. The VehiclePool was being iterated twice. Do it only once and assume only primary vehicles make use of last_station_visited and last_loading_station.

SamuXarick committed:
Codechange: Iterate group vehicle list for RemoveRoadStop, RemoveAirport and HasStationInUse

SamuXarick committed:
Codechange: Simply check whether the group vehicle list is empty for ToolbarVehicleClick

SamuXarick committed:
Codechange: Iterate train group lists for CheckTrainsLengths

SamuXarick committed:
Codechange: Iterate train group list for CcBuildWagon

SamuXarick committed:
Codechange: Iterate group vehicle lists for IntervalTimer<TimerGameCalendar> _vehicles_yearly
